src/operation/buffer/BufferOp.cpp

```diff
@@ -281,6 +284,61 @@ BufferOp::bufferFixedPrecision(const PrecisionModel& fixedPM)
 #endif
 }
 
+/* public static */
+std::unique_ptr<Geometry>
+BufferOp::bufferByZero(const Geometry* geom, bool isBothOrientations)
+{
+    //--- compute buffer using maximum signed-area orientation
+    std::unique_ptr<Geometry> buf0(geom->buffer(0));
+    if (!isBothOrientations)
+        return buf0;
+
+    //-- compute buffer using minimum signed-area orientation
+    BufferOp op(geom);
+    op.isInvertOrientation = true;
+    std::unique_ptr<Geometry> buf0Inv(op.getResultGeometry(0));
+
+    //-- the buffer results should be non-adjacent, so combining is safe
+    if (buf0->isEmpty()) return buf0Inv;
+    if (buf0Inv->isEmpty()) return buf0;
+
+    std::vector<std::unique_ptr<Geometry>> polys;
+    extractPolygons(buf0.release(), polys);
+    extractPolygons(buf0Inv.release(), polys);
+    if (polys.size() == 1) {
+        std::unique_ptr<Geometry> poly(polys.at(0).release());
+        return poly;
+    }
+    if (polys.size() == 0) {
+        return geom->getFactory()->createMultiPolygon();
+    }
+    return geom->getFactory()->createMultiPolygon(std::move(polys));
+}
+
+/* private static */
+void
+BufferOp::extractPolygons(Geometry* geom, std::vector<std::unique_ptr<geom::Geometry>>& polys)
```

Comment:
This looks duplicated from `CascadedUnion`

Reply:
```static std::unique_ptr<geom::Geometry> restrictToPolygons(std::unique_ptr<geom::Geometry> g);```

It's very similar. Need to be able to do it on both `buf0` and `buf0Inv` to get final result.

Reply:
Well, this is one of those Geometry minor surgery things that crops up all over the place.  It could be generalized out into a common class.  Although sometimes there are slight differences in logic which make it easier to just reproduce the basic code.

Reply:
Perhaps cleaner to use `GeometryExtracter` here.  Although it is more general than needed in this context.

Reply:
Depends on the implementation in GEOS. This one and the CascadedUnion one are careful in extracting subgeometries w/o copying. Since it works I'm inclined to just leave it there, as it matches JTS implementation.